### DISCUSSION TITLE
Multi SPL Token Supported Marketplaces

### DISCUSSION THREAD

**Discussion by kespinola** [holaplex/marketplace#178] [Technical Specifications]
### Background

Currently the marketplace has a 1-1 association of marketplace to Metaplex auction house. An auction house can operate in any spl token but not multiple treasury mints per auction house. If a marketplace wants to support multiple spl tokens then it needs an auction house per spl token.

This is an example marketplace JSON for a 1-1 associated marketplace to to auction house.

```
{
  "meta": {
    "name": "Jungle Cats",
    "description": "The Jungle Cats is a collection of 5585 unique high quality 3D generative Cats."
  },
  "theme": {
    "logo": {
      "url": "https://arweave.net:443/o70X307BOrQv-2FhcW-nOY7cpaOYxbllDE3_5xcQfRk",
      "name": "JC Logo White.png",
      "type": "image/png"
    },
    "banner": {
      "url": "https://arweave.net:443/NtD4V1FQw-jODL1n0F5ZdX5uAKUsN7Au9vDH1jk0zr4",
      "name": "watering.jpeg",
      "type": "image/jpeg"
    }
  },
  "creators": [
    {
      "address": "5ujZ8BBFXpcByNbQdxANxbydxUhz5hVavpapvrEUc2Da"
    },
    {
      "address": "HMp36xN691BjS2jDyTLtviuG5KrXFzjGzjgS8UJytZh8"
    }
  ],
  "subdomain": "junglecats",
  "address": {
    "owner": "str3PujRzRnWS6utXdU8XtCA1jJxUWEAVJKVjGnFU6d",
    "auctionHouse": "EsrVUnwaqmsq8aDyZ3xLf8f5RmpcHL6ym5uTzwCRLqbE",
    "store": "GBEbDdc3KgyTQUUnL6BBgA7cAZ9rPdD5CmVY3xAY6k1P",
    "storeConfig": "23twPCvWSSL52UtqejZAneS7h7Cp4N3fqVuFGzwV4mNv"
  }
}
```

## Proposal

Add a field of auctionHouses to the marketplace JSON spec and mark address.auctionHouse as deprecated. This field will track the auction houses generated for the desired payment tokens for the marketplace.

```
auctionHouses: [{ address }]
```

### Indexing

The array field of auctionHouses is tracked in the `store_auction_houses` table which represents the 1 to many relationship of marketplace to auction houses.

```
                         Table "public.store_auction_houses"
        Column        |         Type          | Collation | Nullable | Default
----------------------+-----------------------+-----------+----------+---------
 store_config_address | character varying(48) |           | not null |
 auction_house_address      | character varying(48) |           | not null |
Indexes:
    "store_creators_pkey" PRIMARY KEY, btree (store_config_address, auction_house_address)
    "store_creators_config_addr_index" btree (store_config_address)
```

### API

Any queries that require the marketplace auction house should either return an array of results (1 per auction house) or accept `auctionHouse: PublicKey!` as a field argument.

### Transactions

When the native mint of `So11111111111111111111111111111111111111112` the associated token accounts for the splt token must be included in listing and sale actions.

**Comment by alchemistgo87** [comment]
## Questions & Suggestions

### Field to set default token?
- We can have this field inside store_config_jsons table
- We can set native Solana as the default token for all the existing rows and enable admin to edit the default token from the edit page.

### Field to store list of allowed tokens?
- We can have another field for allowed tokens inside store_config_jsons table
- Logo/Names of these allowed tokens can be displayed in the header of marketplace and user can have option to change the selected token to change the price and other details of marketplace accordingly.

### Option for Admin to add new token?
- Do we provide admin to select from a list of all available spl tokens?
- Or They can input an spl token address and we run a check for its authenticity before updating store config?

### Do we show all activities and bids in the respective views irrespective of token selected by user?
- I guess it would be better if we show all the activities and bids in all cases.


## Other (Suggested) Updates in Api/Database

### auctionHouse query 
AuctionHouse query can be made a part of queryroot (This will help us directly update marketplace view when user changes the selected token)

As field arguments we can have 
- token mint addresss (optional)
- auction house address (optional)
- subdomain (optional, In this case auction house corresponding to default token will be fetched)

### nfts, nft query
- Add field argument named auction_house
- This field argument will help us fetch the correct required listing to show nft price


### BidReceipt, PurchaseReceipt, ListingReceipt, NftActivity
- Add a new field with name amount to show amount of tokens used/required for transaction
- Add fields to show token name, logo image, address (will be useful to show in the bids and activities list if we decide to show all types of bids/activities irrespective of what token user has selected)
- value of price field should be now calculated by converting price of token to lamports.

**Reply by kespinola** [reply]
Field to set default token?
To start lets assume SOL is the default token. If feature request in the future can add.

Field to store list of allowed tokens?
Need to explore if some provides an API of supported SPL tokens. Looks like solana actually hosts an sdk for looking up tokens.

https://github.com/solana-labs/token-list

> Do we provide admin to select from a list of all available spl tokens?

Yes pick list using token-list npm package.

> Do we show all activities and bids in the respective views irrespective of token selected by user?

The nft activity should display listings for any auction house associated to the marketplace.

> nfts, nft query

Agreed on adding auction_houses as option to scope nfts related to certain auction houses. I'm looking into this some now.

> Add fields to show token name, logo image, address (will be useful to show in the bids and activities list if we decide to show all types of bids/activities irrespective of what token user has selected)

Ideally like to do the token type as a local only field on the auction house. Apollo fetches using the token list sdk. I can help look into this.

**Comment by alchemistgo87** [comment]
### More Suggestions

**1) NFTActivity -** 

Require a new field (mintToken) to know the type of token used for sell/purchase (Or instead of adding mintToken field, we can send full auctionHouse data by updating the type of auctionHouse field already present from String to AuctionHouse)

Similarly for ListingReceipt, PurchaseReceipt, BidReceipt

**2) charts query -** 

OPTION 1:

In result, Instead of sending a single PriceChart send a list of PriceCharts corresponding to each auctionHouse. (Note: For this we will also be required to add an auctionHouse field inside the PriceChart struct)

OPTION 2:

In result send combined Pricechart of all kinds of spl-tokens by converting the spl-token price to lamports (using oracles)
